### ISSUE TITLE
Logging in to GCP using a service account

### ISSUE THREAD

**Issue by bilelomrani1** [skypilot-org/skypilot#2109]
I was thinking about a potential interesting use case for SkyPilot in my team: being able to automatically launch our GPU test suite in CI. This is something we struggled with for quite some time, and I did not find any good solution for this. We recently introduced SkyPilot into our workflow and currently, SkyPilot already helps us launch these tests from a local development machine. But going one step further, it would be fantastic if launching the test suite could be fully automated in our CI pipeline.

For some context, we are using GCP and a Gitlab self-managed instance.

My idea was to interact with the GCP project in a CI runner through a service account, I tried to prototype a working solution, here is what I did:
- creating a dedicated machine called `e2e-test-gpu` in our GCP project using SkyPilot
- creating a service account with a minimal set of permissions as discussed in #2100 and exported a JSON service account key
- adding the key in a `$GPU_TESTING_SERVICE_ACCOUNT_KEY` Variable in Gitlab CI
- creating a `launch-gpu-tests.yml` SkyPilot configuration with a command to launch our GPU test suite
- adding the following job in our `.gitlab-ci.yml`:

```yaml
GPU tests:
  image: google/cloud-sdk:latest
  stage: Test
  before_script:
    - pip install --upgrade pip
    - pip install skypilot[gcp]
  script:
    - echo $GPU_TESTING_SERVICE_ACCOUNT_KEY > gpu-testing-service-account-key.json
    - export GOOGLE_APPLICATION_CREDENTIALS="gpu-testing-service-account-key.json"
    - gcloud auth activate-service-account --key-file gpu-testing-service-account-key.json 
    - gcloud config set project $GCP_PROJECT_ID
    - sky check
    - sky launch -i 1 -c e2e-test-gpu tests/gpu-launcher/launch-gpu-tests.yml
```

In the CI runner, the command `sky check` warns that it failed to connect to GCP:

```
  GCP: disabled          
    Reason: GCP tools are not installed or credentials are not set. Run the following commands:
      $ pip install google-api-python-client
      $ conda install -c conda-forge google-cloud-sdk -y
      $ gcloud init
      $ gcloud auth application-default login
    For more info: https://skypilot.readthedocs.io/en/latest/getting-started/installation.html
```

The dependencies `google-cloud-sdk` and `google-api-python-client` are correctly installed, so it seems that it's the authentification part that caused the failure. My guess is that SkyPilot necessarily expects a user to be logged in, but for automated workflows we probably want to use a service account instead. Is there a technical limitation preventing such a use case, and if not is it a feature of interest to you? Such a feature would unlock interesting use cases.

EDIT: A quick skimming through the code reveals that the file `~/.config/gcloud/application_default_credentials.json` is explicitly checked for existence on [this line](https://github.com/skypilot-org/skypilot/blob/d6c2b7a0748b2e1b935a6bace2e3f76f3f600f79/sky/clouds/gcp.py#L488) and may be the culprit here. This file is a user credential file and does not exist when logged through a service account. Instead, the standard practice is to set the `$GOOGLE_APPLICATION_CREDENTIALS` environment variable to the key's path, so we can check if this variable is set before using the user credential.

**Comment by Michaelvll** [comment]
Thanks for reporting this issue @bilelomrani1! We are very excited to see the potential of this CI use case.
To fix the issue, as you mentioned, we can probably fallback to the environment variable for finding the credential file in the SkyPilot code.

As a temporary workaround to unblock your use case, would you like to try changing the first command in your script to the following and use the `~/.config/gcloud/application_default_credentials.json` in the later commands:
```
mkdir -p ~/.config/gcloud
echo $GPU_TESTING_SERVICE_ACCOUNT_KEY > ~/.config/gcloud/application_default_credentials.json
```

I tried the above locally and it seems working for me. : )

**Comment by bilelomrani1** [comment]
Hi @Michaelvll, in fact I was just about to send some updates about this, I did fix it with an even simpler trick: I just created an empty `application_default_credentials.json`:

```yaml
GPU tests:
  image: google/cloud-sdk:latest
  stage: Test
  before_script:
    - apt-get install rsync -y
    - pip install --upgrade pip
    - pip install skypilot[gcp]
    - echo $GPU_TESTING_SERVICE_ACCOUNT_KEY > gpu-testing-service-account-key.json
    - export GOOGLE_APPLICATION_CREDENTIALS="gpu-testing-service-account-key.json"
    - gcloud auth activate-service-account --key-file gpu-testing-service-account-key.json 
    - gcloud config set project $GCP_PROJECT_ID
    # Hack to bypass SkyPilot hardcoded verification
    # See https://github.com/skypilot-org/skypilot/issues/2109
    - touch ~/.config/gcloud/application_default_credentials.json
  script:
    - sky launch -i 2 -y -c e2e-test-gpu tests/gpu-launcher/launch-gpu-tests.yml
    - rsync gcpuser@e2e-test-gpu:/home/gcpuser/sky_workdir/.coverage.gpu_tests .coverage.gpu_tests
  after_script:
    # `after_script` does not share the same shell as `before_script` and `script`
    - echo $GPU_TESTING_SERVICE_ACCOUNT_KEY > gpu-testing-service-account-key.json
    - export GOOGLE_APPLICATION_CREDENTIALS="gpu-testing-service-account-key.json"
    - gcloud auth activate-service-account --key-file gpu-testing-service-account-key.json 
    - gcloud config set project $GCP_PROJECT_ID
    # Hack to bypass SkyPilot hardcoded verification
    # See https://github.com/skypilot-org/skypilot/issues/2109
    - touch ~/.config/gcloud/application_default_credentials.json
    - sky stop -y e2e-test-gpu
  artifacts:
    paths:
      - .coverage.gpu_tests
    expire_in: 1 hour
    when: always
```

It took me a few minutes to realize but I also needed to install `rsync`, but I saw after the fact that you mentioned this in #2113, so we are on same page!

After some testing, I can confidently say that it works exactly as I intended, and it was not that hard to pull out!

**Comment by Michaelvll** [comment]
Hey @bilelomrani1, we just fixed this hardcoded file in our code #2116, this should now work without the `touch ~/.config/gcloud/application_default_credentials.json` hack. Please feel free to have a try with the latest master branch. : )
